File: dashboard/src/app/api/models/puzzle-hero.ts
import { Question, QuestionTypes } from "./question";
import { Team } from "./team";

export enum TrackTypes {
    Crypto = "crypto",
    Gaming = "gaming",
    Scavenger = "scavenger",
    Sponsor = "sponsor",
    Vine = "vine",
    EsotericLanguages = "esoteric_languages",
    DataAnalysis = "data_analysis",
    Recon = "recon",
    Steganography = "steganography",
    Debugging = "debugging",
    Forensics = "forensics",
    Charades = "charades",
    ReverseEngineering = "reverse_engineering",
    CodeGolf = "code_golf",
    Regex = "regex",
    Scripting = "scripting",
    Food = "food",
    Karaoke = "karaoke",
    Pwning = "pwning",
    Authentication = "authentication",
    StaticAnalysis = "static_analysis",
    ArtificialIntelligence = "artificial_intelligence",
    DefensiveSecurity = "defensive_security",
    OffensiveSecurity = "offensive_security"
}

export interface PuzzleInfo {
    id: string;
    label?: string;
    type?: QuestionTypes;
    completed?: boolean;
    locked?: boolean;
    description?: { [lang: string]: string };
    dependsOn?: string;
    score?: number;
    question?: Question;
    answers?: Answers[];
    answersCount?: number;
}

export interface Track {
    _id: string;
    label: string;
    type: TrackTypes;
    puzzles: PuzzleInfo[];
    releaseDate: string;
    endDate: string;
}

export interface Answers {
    _id: string;
    timestamp: string | Date;
    puzzle: string;
    teamId: Team;
}

export interface PuzzleHero {
    tracks?: Track[];
    open: boolean;
    releaseDate: string | Date;
    endDate: string | Date;
    scoreboardEndDate: string | Date;
}

export interface PuzzleHeroInfo {
    open: boolean;
    scoreboardOpen: boolean;
}

export interface Score {
    teamId: string;
    teamName: string;
    schoolName: string;
    score: number;
}

export interface TeamSeries {
    name: string;
    series: { value: number, name: Date }[];
}

File: mobile/lib/domain/puzzle-hero.dart
class PuzzleInfo {
    String id;
    String label;
    Map<String, dynamic> description;
    bool completed;
    bool locked;

    PuzzleInfo({
        this.label,
        this.completed,
        this.locked,
        this.description
    });

    PuzzleInfo.fromMap(Map<String, dynamic> map) {
        id = map["_id"];
        label = map["label"];
        description = Map.castFrom<String, dynamic, String, String>(map['description'] ?? {});
        completed = map["completed"];
        locked = map["locked"];
    }
}

class Track {
    String id;
    String label;
    List<PuzzleInfo> puzzles;
    DateTime releaseDate;
    DateTime endDate;

    Track({
        this.id,
        this.label,
        this.puzzles,
        this.releaseDate,
        this.endDate
    });

    Track.fromMap(Map<String, dynamic> map) {
        id = map["_id"];
        label = map["label"];
        puzzles = List.castFrom<dynamic, PuzzleInfo>(map["puzzles"].map((puzzle) => PuzzleInfo.fromMap(puzzle)).toList());
        releaseDate = DateTime.parse(map["releaseDate"]);
        endDate = DateTime.parse(map["endDate"]);
    }
}

class PuzzleHero {
    List<Track> tracks;
    DateTime releaseDate;
    DateTime endDate;

    PuzzleHero({
        this.tracks,
        this.releaseDate,
        this.endDate
    });

    PuzzleHero.fromMap(Map<String, dynamic> map) {
        tracks = List.castFrom<dynamic, Track>(map["tracks"].map((track) => Track.fromMap(track)).toList());
        releaseDate = DateTime.parse(map["releaseDate"]);
        endDate = DateTime.parse(map["endDate"]);
    }
}
